docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
+
+  interface Observer {
+    /**
+     * Invoked when changes are made to the storage.
+     */
+    fun onStorageUpdated()
+  }
+}
+
+// Already exists in browser-storage-sync.
+interface PlacesHistoryStorage {
+  fun getTopFrecentSites(num: Int)
+}
+
+class DefaultTopSitesStorage(
+  val pinnedSitesStorage: PinnedSitesStorage,
+  val historyStorage: HistoryStorage? = null
+) : TopSitesStorage {
+
+  /**
+   * Merge data sources here, return a single list of top sites.
+   */
+  override fun getTopSites(numFrecent): List<TopSite>
+}
+
+/**
+ * Use cases can be used for adding a top from different places like a context menu.
+ */
+class TopSitesUseCases(topSitesStorage: TopSitesStorage) {
+  val addTopSites: AddTopSiteUseCase
+  val removeTopSites: RemoveTopSiteUseCase
+}
+/**
+ * View-bound feature that updates the UI when changes are made.
+ */
+class TopSitesFeature(
+  val storage: TopSitesStorage,
+  val presenter: TopSitesPresenter,
+  val view: TopSitesView,
+  val defaultSites: () -> List<TopSites>
+) {
+  fun start()
+  fun stop()
+}
+```
+
+## Drawbacks
+
+* `TopSitesStorage` always pulls sites from a persistent storage (Room). This could cause an excessive amount of reads from disk - this is drawback that exists in today's implementation as well. Places has an in-memory cache that should not be affected by this.
+
+## Rationale and alternatives
+
+- This will remove our dependency on `LiveData` in Fenix by updating the `TopSitesView` when we have the top sites available to display; see [#7459](https://github.com/mozilla-mobile/android-components/issues/7459).
+- We can reduce the multiple implementations of the TopSitesFeature by introducing it into Android Components.
+
+## Prior art
+
+* Fenix abstraction of Top Sites: [TopSiteStorage.kt](https://github.com/mozilla-mobile/fenix/blob/3d3153039ce794df09a243968b888ae7cb856d9b/app/src/main/java/org/mozilla/fenix/components/TopSiteStorage.kt)
+* The RFC structure can be see in other components such as, [FindInPageFeature](https://github.com/mozilla-mobile/android-components/blob/master/components/feature/findinpage/src/main/java/mozilla/components/feature/findinpage/FindInPageFeature.kt)  & [SyncedTabsFeature](https://github.com/mozilla-mobile/android-components/blob/master/components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/SyncedTabsFeature.kt).
+
+## Unresolved questions
+
+* When the user removes a top site from the UI that comes from `TopSitesStorage`, we remove it from the storage. When a top site comes from frecent, what do we do? Should we be removing the frecent result?
```

Comment:
Hm would it be within scope to make an exclusions list of sites the user has removed and presumably doesn't want to see as a top site? I think this would be a fairly common use case 😬

Reply:
@gabrielluong mentioned this too! The implementation of a blacklist could definitely work at the `TopSitesStorage`, where does this blacklist get stored is something we should figure out.

With the frecent data coming from the history storage, adding an exclusion list would mean removing sites that show up from there. This will _add_ to our storage layer such that our exclusion list could get much larger over time as we keep asking for the next frecent site and moving it to our blacklist.

I was hoping @grigoryk had some ideas what we should could do here.

Reply:
I also realized we don't have an API from A-S to be able to do this today. :\

A hacky implementation would be based on our exclusion list size which assumes our exclusion list matches the frecent data:

```kotlin
// Alter request to get actual new frecent results
val requestNumOfFrecent = exclusionList.size + originalFrecentRequest size.

// Get a larger list of frecents than what we want
val frecentList = historyStorage.getTopFrecentSites(requestNumOfFrecent)

// Throw out the older frecents because we blacklisted them from previous attempts
val actualFrecentList = frecentList.take(exclusionList.size, frecentList.size) 
```

Reply:
Yeah and good to keep in mind that the user will probably need to be able to edit or at least reset that list as well a la Site Permissions, Tracking Protection, Logins exceptions 😕

Reply:
@ekager i recall that in fennec's contextual menus, "remove" was _by far_ the most popular option :)

I think what makes most sense at this level is to just assume that the frecency API will not return anything we don't want. Then you don't need to worry about counts and stuff.

Then, we'll need an exclusion API (`ignoreFrecentSite`, `resetIgnoredFrecentSites`). This could be entirely implemented within `PlacesHistoryStorage` on top of the A-S API, or within A-S itself. Or we do the A-C impl first, then move it to A-S if needed... I imagine iOS will have an exact similar flow here, and will benefit from this API.

Reply:
cc @linacambridge to weigh in on the exclusion API.

Reply:
Regarding the exclusion API, @ekager and Vesta also noted that Fennec and Desktop provide a "Remove from history" option in the context menu for frecent top sites.

Reply:
From what I can see - Desktop has "Dismiss" and "Remove from History" on their suggestions from history on new tab page. I assume we would _want_ to at least think about these two options and come up with something flexible enough to handle them. I'm worried if we don't add something to handle this case now, it's going to be way harder to add later.

Reply:
We should absolutely handle this case. Again, in Fennec "remove a top site" was a very, very popular action.

Reply:
In Fennec - what did "Remove a top site" do though?

Reply:
@gabrielluong Yep, see my comment above—I wonder if we can push both pinning and ignoring into a-s. It feels inconsistent to have Places manage the "never show this site in top sites" state, but a-c (and eventually iOS, and Desktop) manage the pinned ("always show this site in top sites") state. And then the a-s query would have enough information to do the right thing, where you wouldn't have to request `exclusionList.size + originalFrecentRequest.size` sites and then filter out the excluded ones.